src/modules/uuv_att_control/uuv_att_control.cpp

```diff
@@ -232,155 +187,82 @@ void UUVAttitudeControl::control_attitude_geo(const vehicle_attitude_s &att, con
 	pitch_u = torques(1);
 	yaw_u = torques(2);
 
-	//Quatf q_att(_att.q[0], _att.q[1], _att.q[2], _att.q[3]);
-	//Matrix3f _rot_att = q_att.to_dcm();
-	/** Vector3f current_velocity_boat;
-
-	current_velocity_boat(0) = _local_pos.vx;
-	current_velocity_boat(1) = _local_pos.vy;
-	current_velocity_boat(2) = _local_pos.vz;
-
-	current_velocity_boat = q_att.to_dcm() * current_velocity_boat;
-	*/
-
 	// take thrust as
-	thrust_u = _vehicle_attitude_sp.thrust_body[0];
+	thrust_u = attitude_setpoint.thrust_body[0];
 
 	constrain_actuator_commands(roll_u, pitch_u, yaw_u, thrust_u);
 	/* Geometric Controller END*/
 }
 
-
-void UUVAttitudeControl::run()
+void UUVAttitudeControl::Run()
 {
-	_vehicle_attitude_sp_sub = orb_subscribe(ORB_ID(vehicle_attitude_setpoint));
-	_vehicle_attitude_sub = orb_subscribe(ORB_ID(vehicle_attitude));
-	_angular_velocity_sub = orb_subscribe(ORB_ID(vehicle_angular_velocity));
-	_local_pos_sub = orb_subscribe(ORB_ID(vehicle_local_position));
-	_vcontrol_mode_sub = orb_subscribe(ORB_ID(vehicle_control_mode));
-
-	_manual_control_setpoint_sub = orb_subscribe(ORB_ID(manual_control_setpoint));
-
-	_sensor_combined_sub = orb_subscribe(ORB_ID(sensor_combined));
-
-
-	/* rate limit control mode updates to 5Hz */
-	orb_set_interval(_vcontrol_mode_sub, 200);
-
-	parameters_update(true);
+	if (should_exit()) {
+		_vehicle_attitude_sub.unregisterCallback();
+		exit_and_cleanup();
+		return;
+	}
 
-	/* wakeup source(s) */
-	px4_pollfd_struct_t fds[3];
+	/* check vehicle control mode for changes to publication state */
+	_vcontrol_mode_sub.update(&_vcontrol_mode);
 
-	/* Setup of loop */
-	fds[0].fd = _vehicle_attitude_sub;
-	fds[0].events = POLLIN;
-	fds[1].fd = _manual_control_setpoint_sub;
-	fds[1].events = POLLIN;
-	fds[2].fd = _sensor_combined_sub;
-	fds[2].events = POLLIN;
+	/* update parameters from storage */
+	parameters_update();
 
+	vehicle_attitude_s attitude;
 
-	while (!should_exit()) {
-		/* wait for up to 500ms for data */
-		int pret = px4_poll(&fds[0], (sizeof(fds) / sizeof(fds[0])), 500);
+	/* only run controller if attitude changed */
+	if (_vehicle_attitude_sub.update(&attitude)) {
+		vehicle_angular_velocity_s angular_velocity;
```

Comment:
The system would have to be quite broken for this to not be valid, but to be paranoid you can initialize the struct first.
```suggestion
		vehicle_angular_velocity_s angular_velocity{};
```

Reply:
I'll do, thanks for the advice.